File: cpp/CurrentLimits/src/main/cpp/Robot.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "Robot.h"
#include <iostream>

using namespace ctre::phoenix6;

void Robot::RobotInit() {
  /* Configure the Talon to use a supply limit of 1 amps IF we exceed 4 amps for over 1 second */
  configs::TalonFXConfiguration toConfigure{};
  m_currentLimits.SupplyCurrentLimit = 1; // Limit to 1 amps
  m_currentLimits.SupplyCurrentThreshold = 4; // If we exceed 4 amps
  m_currentLimits.SupplyTimeThreshold = 1.0; // For at least 1 second
  m_currentLimits.SupplyCurrentLimitEnable = true; // And enable it

  m_currentLimits.StatorCurrentLimit = 20; // Limit stator to 20 amps
  m_currentLimits.StatorCurrentLimitEnable = true; // And enable it

  toConfigure.CurrentLimits = m_currentLimits;

  m_fx.GetConfigurator().Apply(toConfigure);
}
void Robot::RobotPeriodic() {
  /* Tie output of joystick to output of motor for current limit testing */
  m_fx.SetControl(m_output.WithOutput(m_joystick.GetLeftY()));

  if(m_joystick.GetAButtonPressed()) {
    /* Toggle the supply limit enable */
    m_currentLimits.SupplyCurrentLimitEnable ^= true;
    std::cout << "Setting supply limit to " << m_currentLimits.SupplyCurrentLimitEnable << std::endl;
    m_fx.GetConfigurator().Apply(m_currentLimits);
  }
  if(m_joystick.GetBButtonPressed()) {
    /* Toggle the stator limit enable */
    m_currentLimits.StatorCurrentLimitEnable ^= true;
    std::cout << "Setting stator limit to " << m_currentLimits.StatorCurrentLimitEnable << std::endl;
    m_fx.GetConfigurator().Apply(m_currentLimits);
  }
    if(printCount++ > 20) {
      printCount= 0;
      std::cout << "Supply current: " << m_fx.GetSupplyCurrent() << std::endl;
      std::cout << "Stator current: " << m_fx.GetStatorCurrent() << std::endl;
    }
}

void Robot::AutonomousInit() {}
void Robot::AutonomousPeriodic() {}

void Robot::TeleopInit() {}
void Robot::TeleopPeriodic() {}

void Robot::DisabledInit() {}
void Robot::DisabledPeriodic() {}

void Robot::TestInit() {}
void Robot::TestPeriodic() {}

void Robot::SimulationInit() {}
void Robot::SimulationPeriodic() {}

#ifndef RUNNING_FRC_TESTS
int main() {
  return frc::StartRobot<Robot>();
}
#endif

File: MotionMagic/src/main/java/frc/robot/Robot.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot;

import org.littletonrobotics.junction.LoggedRobot;

import com.ctre.phoenix6.StatusCode;
import com.ctre.phoenix6.configs.FeedbackConfigs;
import com.ctre.phoenix6.configs.MotionMagicConfigs;
import com.ctre.phoenix6.configs.Slot0Configs;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.controls.MotionMagicVoltage;
import com.ctre.phoenix6.hardware.TalonFX;

import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.XboxController;
import frc.robot.sim.PhysicsSim;

/**
 * The VM is configured to automatically run this class, and to call the functions corresponding to
 * each mode, as described in the TimedRobot documentation. If you change the name of this class or
 * the package after creating this project, you must also update the build.gradle file in the
 * project.
 */
public class Robot extends LoggedRobot {
  private final TalonFX m_fx = new TalonFX(1, "canivore");
  private final MotionMagicVoltage m_mmReq = new MotionMagicVoltage(0);
  private final XboxController m_joystick = new XboxController(0);

  private int m_printCount = 0;

  private final Mechanisms m_mechanisms = new Mechanisms();

  @Override
  public void simulationInit() {
    PhysicsSim.getInstance().addTalonFX(m_fx, 0.001);
  }

  @Override
  public void simulationPeriodic() {
    PhysicsSim.getInstance().run();
  }

  /**
   * This function is run when the robot is first started up and should be used for any
   * initialization code.
   */
  @Override
  public void robotInit() {
    TalonFXConfiguration cfg = new TalonFXConfiguration();

    /* Configure current limits */
    MotionMagicConfigs mm = cfg.MotionMagic;
    mm.MotionMagicCruiseVelocity = 60; // 5 rotations per second cruise
    mm.MotionMagicAcceleration = 325; // Take approximately 0.5 seconds to reach max vel
    // Take approximately 0.2 seconds to reach max accel 
    mm.MotionMagicJerk = 0;

    Slot0Configs slot0 = cfg.Slot0;
    slot0.kP = 0;
    slot0.kI = 0;
    slot0.kD = 0;
    slot0.kA = 0.0202;
    slot0.kV = 0.109;
    slot0.kS = 0.2; // Approximately 0.25V to get the mechanism moving

    FeedbackConfigs fdb = cfg.Feedback;
    fdb.SensorToMechanismRatio = 12.8; //6.75:1

    StatusCode status = StatusCode.StatusCodeNotInitialized;
    for(int i = 0; i < 5; ++i) {
      status = m_fx.getConfigurator().apply(cfg);
      if (status.isOK()) break;
    }
    if (!status.isOK()) {
      System.out.println("Could not configure device. Error: " + status.toString());
    }
  }

  @Override
  public void robotPeriodic() {
    if (m_printCount++ > 10) {
      m_printCount = 0;
      System.out.println("Pos: " + m_fx.getPosition());
      System.out.println("Vel: " + m_fx.getVelocity());
      System.out.println();
    }
    m_mechanisms.update(m_fx.getPosition(), m_fx.getVelocity());
  }

  @Override
  public void autonomousInit() {}

  @Override
  public void autonomousPeriodic() {}

  @Override
  public void teleopInit() {}

  @Override
  public void teleopPeriodic() {
    /* Deadband the joystick */
    double leftY = m_joystick.getLeftY();
    if(leftY > -0.1 && leftY < 0.1) leftY = 0;

    m_fx.setControl(m_mmReq.withPosition(leftY * 10).withSlot(0));
    if(m_joystick.getBButton()) {
      m_fx.setPosition(1);
    }
  }

  @Override
  public void disabledInit() {}

  @Override
  public void disabledPeriodic() {}

  @Override
  public void testInit() {}

  @Override
  public void testPeriodic() {}
}
